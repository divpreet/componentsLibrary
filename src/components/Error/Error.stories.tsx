import * as React from 'react';
import { Story } from '@storybook/react';
import { withKnobs } from '@storybook/addon-knobs';
import Error, { ErrorProps } from './Error';

export default {
    title: 'Surfaces / Error',
    component: Error,
    componentSubtitle: 'Displays an Error page',
    decorators: [withKnobs]
};

const props: ErrorProps = {
    heading: 'Test error heading',
    message: 'Test error message',
    secondaryMessage: 'Test secondary message',
    resetBtn: {
        text: 'Reset text',
        iconClass: 'icon-back',
        onClick: () => {
            console.log('test');
        }
    }
};

const getThemedComponent = (args?) => {
    return <Error {...args} />;
};

export const component: Story = () => {
    return getThemedComponent();
};

export const WithAllProps: Story = () => {
    const properties = { ...props };
    return getThemedComponent(properties);
};

export const SmallVariant: Story = () => {
    const variant: 'small' | 'standard' = 'small';
    const properties = { ...props, variant };
    return getThemedComponent(properties);
};

export const WithoutHeading: Story = () => {
    const properties = { ...props };
    properties.heading = undefined;
    return getThemedComponent(properties);
};

export const WithoutMessage: Story = () => {
    const properties = { ...props };
    properties.message = undefined;
    return getThemedComponent(properties);
};

export const WithoutSecondaryMessage: Story = () => {
    const properties = { ...props };
    properties.secondaryMessage = undefined;
    return getThemedComponent(properties);
};

export const WithoutResetBtn: Story = () => {
    const properties = { ...props };
    properties.resetBtn = undefined;
    return getThemedComponent(properties);
};

export const WithoutResetBtnText: Story = () => {
    const properties = { ...props };
    properties.resetBtn.text = undefined;
    return getThemedComponent(properties);
};

export const WithoutResetBtnIcon: Story = () => {
    const properties = { ...props };
    properties.resetBtn.iconClass = undefined;
    return getThemedComponent(properties);
};
