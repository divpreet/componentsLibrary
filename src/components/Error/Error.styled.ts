import { dclStyled } from '../Theme';
import * as sizes from '../../styles/dclSizes.json';

const ErrorStyle = dclStyled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    height: 100%;
    width: 100%;
    padding: ${sizes.spacing.factor4.s4};
    box-sizing: border-box;

    .dclError__content {
        display: flex;
        flex-direction: column;
        height: auto;
        text-align: center;
    }
    .dclError__body {
        flex: auto;
        display: flex;
        flex-direction: column;
        justify-content: center;
        padding: ${sizes.spacing.factor4.s4};
    }
    .dclError__title {
        flex: none;
        color: ${({ theme }) => theme.text.primary};
        font-size: ${sizes.typography.size.xxlarge};
        font-weight: ${sizes.typography.weight.bold};
        line-height: 1.2em;
        margin: ${sizes.spacing.factor4.s6} 0;
    }
    .dclError__message {
        font-size: ${sizes.typography.size.large};
        font-weight: 400;
        color: ${({ theme }) => theme.text.secondary};
        margin-top: 0;
        margin-bottom: ${sizes.spacing.factor4.s6};
    }
    .dclError__secondary {
        font-size: ${sizes.typography.size.medium};
        font-weight: 400;
        color: ${({ theme }) => theme.text.secondary};
    }
    .dclError__resetBtn {
        margin: 0 auto;
    }
    .dclError__resetIcon {
        font-size: ${sizes.typography.size.small};
    }
    .dclError__logo {
        height: ${sizes.spacing.factor4.s87};
        width: ${sizes.spacing.factor4.s100};
        margin: ${sizes.spacing.factor4.s4} auto;
    }
    &.dclError__small {
        .dclError__title {
            font-weight: 400;
        }
        .dclError__logo {
            height: ${sizes.spacing.factor4.s40};
            width: ${sizes.spacing.factor4.s40};
        }
    }
`;

export default ErrorStyle;
