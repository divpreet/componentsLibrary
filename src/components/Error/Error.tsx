import * as React from 'react';
import clsx from 'clsx';
import Button from '@material-ui/core/Button';
import ErrorStyled from './Error.styled';
import RestrictedAccess from '../../assets/restrictedAccess.svg';

export interface ErrorProps {
    className?: string;
    variant?: 'standard' | 'small';
    heading?: string;
    message?: string;
    secondaryMessage?: string;
    resetBtn?: {
        text?: string;
        iconClass?: string;
        onClick: () => void;
    };
}

const Error: React.FC<ErrorProps> = ({
    className,
    heading = 'Error',
    message = 'An error occurred.',
    secondaryMessage,
    resetBtn,
    variant = 'standard'
}) => {
    const renderMessage = (msg: string) => ({ __html: msg });
    // eslint-disable-next-line camelcase
    const customClassName = clsx(className, { dclError__small: variant === 'small' });

    const ResetBtn = () => (
        <Button className="dclError__resetBtn" onClick={resetBtn.onClick}>
            {resetBtn.iconClass && (
                <span className={`icon ${resetBtn.iconClass} dclError__resetIcon`} />
            )}
            {resetBtn.text || 'Reset'}
        </Button>
    );

    const Logo = () => <img className="dclError__logo" src={RestrictedAccess} alt="Error logo" />;

    return (
        <ErrorStyled className={customClassName}>
            <div className="dclError__content">
                <div className="dclError__body">
                    {variant === 'small' && <Logo />}
                    <h1
                        className="dclError__title"
                        dangerouslySetInnerHTML={renderMessage(heading)}
                    />
                    <h2
                        className="dclError__message"
                        dangerouslySetInnerHTML={renderMessage(message)}
                    />
                    {variant === 'standard' && <Logo />}
                    {resetBtn && <ResetBtn />}
                    {secondaryMessage && (
                        <h3
                            className="dclError__secondary"
                            dangerouslySetInnerHTML={renderMessage(secondaryMessage)}
                        />
                    )}
                </div>
            </div>
        </ErrorStyled>
    );
};

export default Error;
