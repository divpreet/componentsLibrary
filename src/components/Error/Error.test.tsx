import * as React from 'react';
import Error, { ErrorProps } from './Error';
import { mount, shallow } from 'enzyme';

let props: ErrorProps;

describe('Error', () => {
    beforeEach(() => {
        props = {
            heading: 'Test error heading',
            message: 'Test error message',
            secondaryMessage: 'Test secondary message',
            resetBtn: {
                text: 'Reset text',
                iconClass: 'icon-back',
                onClick: jest.fn()
            }
        };
    });

    it('renders correctly', () => {
        const component = shallow(<Error {...props} />).dive();
        expect(component).toMatchSnapshot();
    });

    it('should display error heading, message, logo, secondary message and button', () => {
        const component = mount(<Error {...props} />);

        expect(component.find('h1').text()).toContain('Test error heading');
        expect(component.find('h2').text()).toContain('Test error message');
        expect(component.find('h3').text()).toContain('Test secondary message');
        const resetBtn = component.find('button');
        expect(resetBtn.text()).toContain('Reset text');
        expect(resetBtn.find('.icon-back').length).toBe(1);
        resetBtn.simulate('click');
        expect(props.resetBtn.onClick).toHaveBeenCalled();
    });

    it('should display default error heading, message, logo if not available', () => {
        const component = mount(<Error />);

        expect(component.find('h1').text()).toContain('Error');
        expect(component.find('h2').text()).toContain('An error occurred.');
        expect(component.find('h3').length).toBe(0);
        expect(component.find('button').length).toBe(0);
    });

    it('should display default reset btn text if not available', () => {
        const properties = { ...props, resetBtn: { onClick: jest.fn() } };
        const component = mount(<Error {...properties} />);

        const resetBtn = component.find('button');
        expect(resetBtn.text()).toContain('Reset');
        expect(resetBtn.find('.icon').length).toBe(0);
        resetBtn.simulate('click');
        expect(properties.resetBtn.onClick).toHaveBeenCalled();
    });
});
