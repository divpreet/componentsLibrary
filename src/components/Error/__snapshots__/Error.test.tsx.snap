// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 100%;
  width: 100%;
  padding: 16px;
  box-sizing: border-box;
}

.c0 .dclError__content {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: auto;
  text-align: center;
}

.c0 .dclError__body {
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 16px;
}

.c0 .dclError__title {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  color: #32373F;
  font-size: 24px;
  font-weight: 500;
  line-height: 1.2em;
  margin: 24px 0;
}

.c0 .dclError__message {
  font-size: 16px;
  font-weight: 400;
  color: #525965;
  margin-top: 0;
  margin-bottom: 24px;
}

.c0 .dclError__secondary {
  font-size: 14px;
  font-weight: 400;
  color: #525965;
}

.c0 .dclError__resetBtn {
  margin: 0 auto;
}

.c0 .dclError__resetIcon {
  font-size: 12px;
}

.c0 .dclError__logo {
  height: 348px;
  width: 400px;
  margin: 16px auto;
}

.c0.dclError__small .dclError__title {
  font-weight: 400;
}

.c0.dclError__small .dclError__logo {
  height: 160px;
  width: 160px;
}

<div
  className="c0"
>
  <div
    className="dclError__content"
  >
    <div
      className="dclError__body"
    >
      <h1
        className="dclError__title"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Test error heading",
          }
        }
      />
      <h2
        className="dclError__message"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Test error message",
          }
        }
      />
      <Logo />
      <ResetBtn />
      <h3
        className="dclError__secondary"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Test secondary message",
          }
        }
      />
    </div>
  </div>
</div>
`;
